File: code/backend/Odin/src/main/kotlin/pt/isel/odin/service/interfaces/CurricularUnitService.kt
package pt.isel.odin.service.interfaces

import pt.isel.odin.controller.dto.curricularunit.CurricularUnitRequest
import pt.isel.odin.model.CurricularUnit

/**
 * Service for curricular units.
 */
interface CurricularUnitService {

    /**
     * Gets a curricular unit by its id.
     *
     * @param id the curricular unit id.
     */
    fun getById(id: Long): CurricularUnit

    /**
     * Gets all curricular units.
     */
    fun getAll(): List<CurricularUnit>

    /**
     * Saves or Updates a curricular unit.
     *
     * @param curricularUnitRequest the curricular unit to save.
     */
    fun save(curricularUnitRequest: CurricularUnitRequest): CurricularUnit

    /**
     * Deletes a curricular unit by its id.
     *
     * @param id the curricular unit id.
     */
    fun delete(id: Long)
}

File: code/frontend/Tyr/src/services/CurricularUnitService.ts
import { makeApiRequest } from '../axios/apiRequest';

export class CurricularUnitService {

    static async getById(id: number): Promise<Course> {
        return makeApiRequest('get', `/curricularunits/${id}`);
    }

    static async getAll(): Promise<Course[]> {
        return makeApiRequest('get', '/curricularunits');
    }

    static async save(curricularUnitRequest: Course): Promise<Course> {
        return makeApiRequest('post', '/curricularunits/save', curricularUnitRequest);
    }

    static async update(curricularUnitRequest: Course): Promise<Course> {
        return makeApiRequest('put', '/curricularunits/update', curricularUnitRequest);
    }

    static async delete(id: number): Promise<void> {
        return makeApiRequest('delete', `/curricularunits/${id}`);
    }
}